File: Problems/No.2336_Smallest_Number_in_Infinite_Set/smallest_number_in_infinite_set.py
class SmallestInfiniteSet(object):

    def __init__(self):
        self.isInSet = {}
        self.InfiniteSet = []
        for i in range(1, 1001):
            self.isInSet[i] = True
            heapq.heappush(self.InfiniteSet, i)

    def popSmallest(self):
        """
        :rtype: int
        """

        smallest = heapq.heappop(self.InfiniteSet)
        self.isInSet[smallest] = False

        return smallest
        

    def addBack(self, num):
        """
        :type num: int
        :rtype: None
        """

        if not self.isInSet[num]:
            heapq.heappush(self.InfiniteSet, num)
            self.isInSet[num] = True

        
        


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)

File: Problems/No.2336_Smallest_Number_in_Infinite_Set/smallest_number_in_infinite_set.cpp
class SmallestInfiniteSet {

private:
    std::map<unsigned int, bool> isInSet;
    std::priority_queue<unsigned int, std::vector<unsigned int>, std::greater<unsigned int>> InfiniteSet;

public:
    SmallestInfiniteSet() {

        for (unsigned int num=1; num<=1000; num++) {

            InfiniteSet.push(num);
            isInSet[num] = true;

        }

    }
    
    int popSmallest() {
        
        unsigned int smallest = InfiniteSet.top();
        InfiniteSet.pop();
        isInSet[smallest] = false;

        return smallest;

    }
    
    void addBack(int num) {
        
        if (isInSet[num] == false) {

            InfiniteSet.push(num);
            isInSet[num] = true;

        }

    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */
